### ISSUE TITLE
Search takes a long time

### ISSUE THREAD

**Issue by ozozozd** [zed-industries/zed#10280]
### Check for existing issues

- [X] Completed

### Describe the bug / provide steps to reproduce it

Running global search takes over a minute even though the codebase is fairly small.
- text search
- case insensitive
- NOT "whole word"

### Environment

Zed: v0.129.2 (Zed)
OS: macOS 13.6.6
Memory: 16 GiB
Architecture: aarch64

### If applicable, attach your `~/Library/Logs/Zed/Zed.log` file to this issue.

2024-04-08T18:07:42+03:00 [INFO] reload git repository ".git"
2024-04-08T18:08:52+03:00 [INFO] reload git repository ".git"
2024-04-08T18:10:01+03:00 [INFO] reload git repository ".git"
2024-04-08T18:11:11+03:00 [INFO] reload git repository ".git"
2024-04-08T18:12:21+03:00 [INFO] reload git repository ".git"
2024-04-08T18:13:31+03:00 [INFO] reload git repository ".git"
2024-04-08T18:14:39+03:00 [INFO] reload git repository ".git"

**Comment by JosephTLyons** [comment]
Hey @ozozozd, are you able to share the repo in which you are having the slowdown? Hoping it is some open source thing where we can take a look at what might be the issue.

**Comment by ozozozd** [comment]
@JosephTLyons not an open source repo unfortunately.

I am suspecting that some of the language servers are lazily started and causing the delay. I am seeing the same delay in jumps and when I check the log there is usually an LSP start log line.

What's the easiest way to turn off specific LSPs (e.g. Prisma, TypeScript) so I can test my hypothesis?

**Comment by JosephTLyons** [comment]
You should be able to disable your language servers like this:

```json
"language_overrides": {
  "C": {
    "enable_language_server": false
  }
}
```

**Comment by ozozozd** [comment]
Thanks. Will report back whether disabling the LSP for Prisma fixes it.

**Comment by ozozozd** [comment]
Disabling the Prisma LSP didn't fix it.

I had noticed that Zed was downloading some LSP servers. And logs indicated that the same LSPs were started several times.

I deleted Zed and reinstalled it and I haven't experienced a delay in ~2 hours, so I think that fixed it.

Sorry this doesn't give you much info other than "LSPs downloaded during upgrades may end up in a bad state causing long delays for saving and jumps"

**Comment by ozozozd** [comment]
Never mind. The delays are back making Zed quite unusable. I think I'll go download a version from a month ago or so.

Last few log lines from a long delay while saving:

```2024-04-09T13:50:21+03:00 [ERROR] crates/project/src/project.rs:3139: no worktree found for diagnostics path "/<my user dir>/Library/Application Support/Zed/languages/pyright/node_modules/pyright/dist/typeshed-fallback/stdlib/builtins.pyi"
2024-04-09T13:50:21+03:00 [INFO] Language server with id 2 sent unhandled notification eslint/noLibrary:
{
  "source": {
    "uri": "<my repo path>/index.ts"
  }
}
2024-04-09T13:50:23+03:00 [INFO] Language server with id 3 sent unhandled notification client/unregisterCapability:
{
  "unregisterations": [
    {
      "id": "9a643c26-55eb-4703-bffb-33e06ae76f24",
      "method": "workspace/didChangeWatchedFiles"
    }
  ]
}
2024-04-09T13:50:50+03:00 [INFO] reload git repository ".git"
```

The reference to `crates/project/src/project.rs:3139` is curious. I am guessing this is a Zed internal because my project doesn't have any Rust.

**Comment by ozozozd** [comment]
968 lines of the following in the logs after a slow search:

```
2024-04-09T14:22:18+03:00 [ERROR] crates/project/src/project.rs:6366: stream did not contain valid UTF-8
```

**Comment by JosephTLyons** [comment]
Appreciate help in trying to debug this @ozozozd, I'll tag in @osiewicz, who has worked on optimizing search in the past - not sure we have enough info...

**Comment by osiewicz** [comment]
That log line is indeed telling a bit:
Do you have `Toggle Include Ignored` option enabled? It looks like we're scanning your Zed install directory:
`/<my user dir>/Library/Application Support/Zed/languages/pyright/node_modules/pyright/dist/typeshed-fallback/stdlib/builtins.pyi` which doesn't sound right.
It is available under filters:
![image](https://github.com/zed-industries/zed/assets/24362066/ba4ce953-e6e7-459b-8d9a-f3acce4f7477)

**Comment by osiewicz** [comment]
It seems like new search implementation is slower in general than what we've had before. I've tested with 0.125 stable and a `timer` query on `qmk_firmware` repository.

**Comment by ozozozd** [comment]
`Toggle Include Ignored` was not enabled. In fact, I just found out about that option.

My git repo had to be garbage collected and/or pruned. I am wondering if it had to do with that. AFAICT Zed doesn't have many git features but maybe there is stuff that isn't exposed in the app and my git repo being bloated affected it?

Also, this is the second time I caught Zed to be the almost biggest consumer of CPU time. The previous instance had Zed above `kernel_task` which was really surprising.

![Screenshot 2024-04-12 at 10 19 01 AM](https://github.com/zed-industries/zed/assets/1986349/2a6afb98-40e9-431a-9384-d90d70576704)

**Comment by ozozozd** [comment]
I installed an older version, upgraded from there to `0.129.2` and I don't have the delays anymore. Unfortunately I can't remember which version I downloaded and the dmg doesn't have the version in the filename.

**Comment by ozozozd** [comment]
Sorry to flood with comments, but I wouldn't assume this is a search specific issue. Search, jump to definition, save share this delay pattern. I would hypothesize that something common to these 3 actions or something that has a lock on a resource these 3 actions wait on is causing the delay.

**Comment by mrnugget** [comment]
> It seems like new search implementation is slower in general than what we've had before. I've tested with 0.125 stable and a `timer` query on `qmk_firmware` repository.

That's interesting. We've made the opening of buffers bounded, to help with unresponsiveness app when searching lots of files, but if it's not a gigantic repository, I wouldn't guess that it's slower.

**Comment by notpeter** [comment]
> I installed an older version, upgraded from there to 0.129.2 and I don't have the delays anymore. Unfortunately I can't remember which version I downloaded and the dmg doesn't have the version in the filename.

@ozozozd It sounds like you managed to get this sorted out after downgrading and upgrading, but I wanted to confirm that you are not still experiencing this issue, and if so close out this issue.

Thanks

**Comment by JosephTLyons** [comment]
I'm going to close this issue out for now. @ozozozd, if you are still hitting this in the most recent version of Zed, please open another issue!
